File: server/chess/pawn.go
package chess

type Pawn struct {
    basePiece
}

func (p *Pawn) String() string {
    if p.Color() == WHITE {
        return "P"
    } else {
        return "p"
    }
}

func (p *Pawn) IsValidMove(b *Board, fromCol, fromRow, toCol, toRow int) (valid bool, execute func()) {
    var dir, startRow int
    if p.Color() == WHITE {
        dir, startRow = 1, 1
    } else {
        dir, startRow = -1, 6
    }

    dCol := toCol - fromCol
    dRow := toRow - fromRow
    target := b[toRow][toCol]

    switch {
    case dCol == 0 && dRow == dir && target == nil:
        // single push
    case dCol == 0 && dRow == 2*dir && fromRow == startRow &&
        target == nil && b[fromRow+dir][fromCol] == nil:
        // double push from starting rank
    case (dCol == 1 || dCol == -1) && dRow == dir &&
        target != nil && target.Color() != p.Color():
        // diagonal capture
    case (dCol == 1 || dCol == -1) && dRow == dir &&
        target == nil && b[fromRow][toCol].MovesMade() == 1:
        // en pessant
        return true, func() {
            b[toRow][toCol] = b[fromRow][fromCol]
            b[fromRow][fromCol] = nil
            b[fromRow][toCol] = nil
        }
    default:
        return false, nil
    }

    return true, func() {
        b[toRow][toCol] = b[fromRow][fromCol]
        b[fromRow][fromCol] = nil
    }
}

File: chess_logic/pawn.cpp
#include "pawn.h"
#include "board.h"
#include <cmath>

bool Pawn::isValidMove(int fromX, int fromY, int toX, int toY, const Board& board) const {
    // White pawns start on row 6 and move toward row 0 (decreasing y).
    // Black pawns start on row 1 and move toward row 7 (increasing y).
    int direction = (color == Color::WHITE) ? -1 : 1;
    int startRow = (color == Color::WHITE) ? 6 : 1;

    int diffX = toX - fromX;
    int diffY = toY - fromY;

    ChessPiece* target = board.getPieceAt(toX, toY);

    if (diffX == 0) {
        // Single move
        if (diffY == direction && target == nullptr) {
            return true;
        }

        // Double move
        if (fromY == startRow && diffY == 2 * direction && target == nullptr) {
            ChessPiece* squareInBetween = board.getPieceAt(fromX, fromY + direction);
            if (squareInBetween == nullptr) {
                return true;
            }
        }

        return false;
    }

    // Diagonal capture
    if (std::abs(diffX) == 1 && diffY == direction) {
        // Normal capture
        if (target != nullptr && target->getColor() != this->color) {
            return true;
        }
        if (target == nullptr && toX == board.getEnPassantTargetX() && toY == board.getEnPassantTargetY()) {
            return true;
        }
    }

    return false;
}
